src/rules/returnUndefinedRule.ts

```diff
@@ -106,27 +103,38 @@ function getReturnKind(node: FunctionLike, checker: ts.TypeChecker): ReturnKind
             return ReturnKind.Void;
         case ts.SyntaxKind.GetAccessor:
             return ReturnKind.Value;
-        default:
     }
 
-    // Go with an explicit type declaration if possible.
-    if (node.type) {
-        return node.type.kind === ts.SyntaxKind.VoidKeyword ? ReturnKind.Void : ReturnKind.Value;
-    }
-
-    const contextualType = node.kind === ts.SyntaxKind.FunctionExpression || node.kind === ts.SyntaxKind.ArrowFunction
-        ? checker.getContextualType(node)
-        : undefined;
+    const contextual = isFunctionExpressionLike(node) ? tryGetReturnType(checker.getContextualType(node)) : undefined;
+    const returnType = contextual !== undefined ? contextual : tryGetReturnType(checker.getTypeAtLocation(node));
 
-    const ty = contextualType || checker.getTypeAtLocation(node);
-    if (!ty) {
-        // Type error somewhere. Bail.
+    if (returnType === undefined) {
         return undefined;
+    } else if (Lint.isTypeFlagSet(returnType, ts.TypeFlags.Void)) {
+        return ReturnKind.Void;
+    } else if (Lint.hasModifier(node.modifiers, ts.SyntaxKind.AsyncKeyword)) {
+        // TODO: Would need access to `checker.getPromisedTypeOfPromise` to do this properly.
```

Comment:
doesn't sound like this implementation is likely to change soon, so let's remove the "TODO" part and just consider this comment an implementation note rather than an expectation of a future change